camel/models/reka_model.py

```diff
@@ -0,0 +1,212 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import os
+from typing import TYPE_CHECKING, Any, Dict, List, Optional
+
+from camel.configs import REKA_API_PARAMS
+from camel.messages import OpenAIMessage
+from camel.models import BaseModelBackend
+from camel.types import ChatCompletion, ModelType
+from camel.utils import (
+    BaseTokenCounter,
+    OpenAITokenCounter,
+    api_keys_required,
+)
+
+if TYPE_CHECKING:
+    from reka.types import ChatMessage, ChatResponse
+
+try:
+    import os
+
+    if os.getenv("AGENTOPS_API_KEY") is not None:
+        from agentops import LLMEvent, record
+    else:
+        raise ImportError
+except (ImportError, AttributeError):
+    LLMEvent = None
+
+
+class RekaModel(BaseModelBackend):
+    r"""Reka API in a unified BaseModelBackend interface."""
+
+    def __init__(
+        self,
+        model_type: ModelType,
+        model_config_dict: Dict[str, Any],
+        api_key: Optional[str] = None,
+        url: Optional[str] = None,
+        token_counter: Optional[BaseTokenCounter] = None,
+    ) -> None:
+        r"""Constructor for Reka backend.
+
+        Args:
+            model_type (ModelType): Model for which a backend is created,
+                one of REKA_* series.
+            model_config_dict (Dict[str, Any]): A dictionary that will
+                be fed into `Reka.chat.create`.
+            api_key (Optional[str]): The API key for authenticating with the
+                Reka service. (default: :obj:`None`)
+            url (Optional[str]): The url to the Reka service.
+            token_counter (Optional[BaseTokenCounter]): Token counter to use
+                for the model. If not provided, `OpenAITokenCounter` will be
+                used.
+        """
+        super().__init__(
+            model_type, model_config_dict, api_key, url, token_counter
+        )
+        self._api_key = api_key or os.environ.get("REKA_API_KEY")
+        self._url = url or os.environ.get("REKA_SERVER_URL")
+
+        from reka.client import Reka
+
+        self._client = Reka(api_key=self._api_key, base_url=self._url)
+        self._token_counter: Optional[BaseTokenCounter] = None
+
+    def _to_openai_response(self, response: 'ChatResponse') -> ChatCompletion:
+        obj = ChatCompletion.construct(
```

Comment:
I see that `ChatCompletion.construct()` is used instead of `ChatCompletion(...)`. Is there a specific reason for this choice?

Reply:
use this method since it's more nature to create ChatCompletion, there's some additional checks in `ChatCompletion.construct()`